### ISSUE TITLE
Should all MathML elements really be potential hyperlinks / match :visited?

### ISSUE THREAD

**Issue by emilio** [mathml-refresh/mathml#125]
https://mathml-refresh.github.io/mathml/#attributes-shared-by-all-mathml-elements says:

> href: Can be used to establish the element as a hyperlink to the specfied URI.

This in Gecko has implications and behaves the same way as `<a>` or `<area>`.

Two things:

 * Is this feature used / desired / worth it? Links are complex.

 * If it is, should it really match `:visited`? I think in WebKit it does not, if I'm reading the code correctly.

**Comment by davidcarlisle** [comment]
certainly the linking is used and worth keeping, in practice they are almost always used with token elements so if that would simplify the implementation could consider restricting href to token elements in core at least.

**Comment by fred-wang** [comment]
Yes, href is necessary otherwise you can't create a link for a subformula, only the whole formula. MathML3 basically allows it on all elements and it was a replacement for the xlink::href attributes: https://www.w3.org/TR/MathML3/chapter6.html#interf.link

We have some stats in https://github.com/mathml-refresh/mathml/issues/55 although we didn't ask this explicitly, probably we should?

For consistency with HTML and SVG, I guess it makes sense to restrict it to only a container element (rather than to all MathML elements), which would be `<mrow>`. Any element can be wrapped in such an `<mrow>` if it's necessary to create a link.

**Comment by davidcarlisle** [comment]
forcing it on mrow would make the mathml a lot more verbose than it already is (in our mathml for example almost every mi is a link) allowing href on all (or most) elements was explicitly raised with whatwg at the time of the mathml integration and the feedback we had at the time was that (as far people were interested in implementing mathml at all) the href attribute wouldn't be a problem.
if it was mrow and token elements (mi,mn,mtext,ms) then that would cover all proactical cases

**Comment by NSoiffer** [comment]
I have all elements, although math, mrow, and leaf elements is a reasonable
compromise if it is hard to implement. For full, a polyfill could add an
mrow in that case.

    Neil

**Comment by fred-wang** [comment]
>     * If it is, should it really match `:visited`? I think in WebKit it does not, if I'm reading the code correctly.

For the record, Gecko has a test for this: https://searchfox.org/mozilla-central/source/layout/style/test/test_visited_reftests.html ; not sure how easy it is to write one with WPT.

**Comment by bkardell** [comment]
After a bunch of discussion at/driven by conversations at our hackfest, I believe (and I think @NSoiffer agrees) that we should limit this to token elements.  I'm not entirely sure that I really understand the real need since `<a href="">` should be legit content of any of the elements and I'm not sure I understand how there is a difference of substance, but I suppose the data seems to suggest that these are where this feature is really used, and that it's mostly (exclusively even?) token elements.  Anyways, the reason is that links have activation behavior and a bunch of specific things that add complexity to the idea of adding something like shadow dom.  I'm not suggesting that adding it will be trivial without or something, but it seems very unlikely to add such features in the near term if we just say that everything can be a link.  

So, for simplicty sake, is there any objection to limiting it?

**Comment by davidcarlisle** [comment]
> So, for simplicty sake, is there any objection to limiting it?

Not really, I have some mrow href... eg the 1..n link in 

https://www.nag.co.uk/numeric/nl/nagdoc_27/flhtml/f08/f08upf.html#jfail

(which is native mathml rendering if you use firefox, mathjax otherwise)

but in the full collection there, there are 30 `<mrow` with href as opposed 97549 `<mi` and 3226 `<mn`, so dropping this is not a big issue for this set of documents and certainly if it causes you any extra implementation issues or difficult discussions with the browser makers then make it token elements only.

**Comment by fred-wang** [comment]
Note that there are existing content and tests using href on mrow.

Technically:
* limiting to mrow would still make all the other cases polyfillable (basically wrap the MathML element to link in an mrow) so it's not making MathML weaker.
* limiting to token elements, one won't be able to make a subset of a formula linkable. Which I believe can be an interesting use cases (e.g. link to a definition of a math notation).

We need to update https://html.spec.whatwg.org/#dom-tabindex to indicate which element should have tabindex 0. Currently, we only test mrow: https://w3c-test.org/mathml/relations/html5-tree/tabindex-001.html

**Comment by fred-wang** [comment]
See also https://github.com/mathml-refresh/mathml/issues/152#issuecomment-582680428

**Comment by NSoiffer** [comment]
The current discussion on #152 really belongs here, so I'm commenting on this issue in the hope the discussion returns to this issue...

There was some discussion about MathML should be like HTML/SVG. SVG does have a `a` element, but according to [this MDN page](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/href), it also allows `href` on 15 elements (with various meanings/restrictions based on the element). That's a long way from "all" elements like the current MathML spec, but it is also a long way from just 4 leaf elements.

In https://github.com/mathml-refresh/mathml/issues/152#issuecomment-583440328, @bzbarsky wrote:
> Also, any time the document base URL changes (e.g pushState), all links have to be invalidated. This means walking the list of all potential links, and if that means all MathML elements that can be a performance issue. There are some things that can be done to optimize this, but they're fragile and complicated.

I can imagine anytime there is a link, there is a much that might need to be updated. What I'm unclear on is why the number of elements that might have a link is the issue as opposed to the number of links. It seems (from my naive point of view), you just need to find the elements that have a link (which in MathML is any element with an href attribute). Why would it make a difference if the number of elements with a _potential_ `href` is 4 or 40? Can you clarify?

**Comment by bzbarsky** [comment]
> it also allows href on 15 elements (with various meanings/restrictions based on the element)

Yes, but those don't act as "links", just like the `src` attribute on HTML's `img` does not act as a "link".

> you just need to find the elements that have a link

This is the "There are some things that can be done to optimize this, but they're fragile and complicated" bit from that comment.  Yes, you can do this, but maintaining that state is a pain and slows down various DOM mutations and whatnot.

**Comment by bkardell** [comment]
Links have all kinds of specialness, so it seems there's a lot of questions that wind up tying into this... The 'linky things that support href' (`a` in both HTML and SVG and `area` in HTML) have a lot more 'stuff' there..  

* They have a default `.tabindex` of 0, even if you don't supply an href - but you can't actually sequentially focus them unless they do.  
* They also mixin `HTMLHyperlinkElementUtils`... we aren't doing that
* They support a bunch more attributes that inform stuff about how to follow the link `target`, `download`, `ping`,  `hreflang`, `type`,  `referrerpolicy` and `rel` with all of the relevant hyperlink annotations:  `noreferrer`, `noopener`, `nofollow`, etc and contain IDL props for some of these things that aren't in a mixin
* they affect selector matching for several CSS pseudo classes - at least some of those have some additional challenges -- `:focus-visible` for example treats some kinds of elements a little specially and doesn't draw a focus ring on them when they are clicked.. links is one of those.  In some places (maybe all now) I think they might not actually get focused even.
* They're never going to be able to (themselves) have a shadow dom probably.

There's a bunch of other tendrils here too I guess - they affect `pushState`, etc, etc - but all of these things are very user facing and at least seem that they need called out or answered at a spec level more.

So, I think there's two fundamental questions here that go to the aims of this initial work:
* How important is it that core (v1) provides an answer to this question that is 100% backward compatible with existing content without a little JavaScript?
* Thusfar, core has been only 'subset' and 'explain existing stuff' - we've not _added_ new things.  Would we be ok with something that broke with that for linking? (ie, add a proper MathML `a` for consistency in the platform and simplicity in tying it all together).. This would render by default as `mrow` and I don't think causes any new sorts of problems but obviously there's not existing content like this, and I'm unsure how it would affect an old system?

**Comment by fred-wang** [comment]
My personal opinion is that ability to link any subset of MathML formulas is an interesting feature. I'm not sure it is as fundamental as other things we are working on though. So I'd be happy to either ignore links for now (with some clever polyfill) OR introduce a new backward-incompatible `<a>` that is clean & consistent with HTML/SVG (this would give a straightforward DOM-converter polyfill).

**Comment by bkardell** [comment]
I have created a relevant pull request for the specification based on our conversation last week - however, in discussing with @fred-wang I think that we agree we'd like to discuss whether it might be a good choice/compromise to make this for token elements natively (most existing content) and think about adding an `<a href>` style thing for other uses... It would be good to make a decision because there are a number of things that follow this decision I think..

**Comment by davidcarlisle** [comment]
Wouldn't the new `<a>`  just be (the existing) `<mrow>`?   For the mathml language design, `<a>` seems slightly strange as all similar mathml elements start with `m` which is a concern, but mainly duplicating mrow seems to be adding complication.

Given how hard you're pushing for this, and re-raising it, I feel I must be missing something?

What does not allowing href on mrow give you? (given that content authors could simply use `<a>` everywhere where they currently use `<mrow>`  whether or not an href is needed).

From compatibility and language design points of view, adding `<a>` seem somewhat unfortunate, but as I say above somewhere, if it helps smooth implementations I don't actually object to it, as smoothing implementations is more or less the point of mathml core.  But (if only for my own education) it would be good to understand what the benefits are.

**Comment by NSoiffer** [comment]
Resolved: all leaf elements and mrow should support href and it's related properties; no other element should.
Rob will review IDL change to support this that Brian created.

Need to write some tests. Brian submitted changes to the core spec.

**Comment by NSoiffer** [comment]
As per the MathML core meeting on 2/320, I was tasked to look into some accessibility issues related to href not being present on an element that can have it. The issue that @bkardell raised is that `tabindex=0` when there is not `href`, but it is not focusable.

I asked a colleague who is very knowledgeable about accessibility and he said he was sure that was not legal. That caused me to look things up and indeed it is legal:
* [Not having href is legal](https://www.w3.org/html/wg/spec/text-level-semantics.html#the-a-element).
* The spec also says setting of the focus flag is _suggested_ for 'a' with href and in Example 5 in this part of the spec about being focusable, it says 'sometimes \<a href=""> (depending on platform conventions)' as one of the focusable nodes. 

I tried out the following example in Chrome and Firefox (I know, it could have been written better): 
```
<h2> a link test </h2>
<p>
<a id="href" href="#a-nohref"> Standard anchor</a>.
<span id="value-href"></span>.
</p>
<p>
<a id="a-nohref"> An anchor, but not linked </a>.
<span id="value"></span>.
</p>
<p>
Here is some stuff outside of 'a'.
</p>
<p>
Here is some math with a link:
<math xmlns="http://www.w3.org/1998/Math/MathML">
<mi id="mi-nohref">x</mi>
</math>
<span id="valueMathNoRef"></span>.
</p>
<p>
Here is some math with a link:
<math xmlns="http://www.w3.org/1998/Math/MathML">
<mi id="mi" href="#a-nohref">x</mi>
</math>
<span id="valueMath"></span>.
</p>
<script>
document.getElementById("value-href").textContent =
    ' tabindex=' + document.getElementById("href").tabIndex +
    ', focus=' + document.getElementById("href").focus;
document.getElementById("value").textContent =
    ' tabindex=' + document.getElementById("a-nohref").tabIndex +
    ', focus=' + document.getElementById("a-nohref").focus;
document.getElementById("valueMathNoRef").textContent =
    ' tabindex=' + document.getElementById("mi-nohref").tabIndex +
    ', focus=' + document.getElementById("mi-nohref").focus;
document.getElementById("valueMath").textContent =
    ' tabindex=' + document.getElementById("mi").tabIndex +
    ', focus=' + document.getElementById("mi").focus;
</script>
```
The results are the same in Chrome and Firefox for the `a` element. In Firefox, for the math/mi you get tabindex=-1 without the href; you get '0' when it is present. However, there is an accessibility warning that no action is associated with it despite the fact that clicking on it does change the focus appropriately. I suspect Brian is saying to himself "that's why we should inherit working code" :-)

I was a bit surprised that NVDA worked as I would want when there is no href (i.e., tabbing would not stop there). I think all screen readers will work the same. In both Firefox and Chrome, the accessibility tree does not show focus in those cases (there is no tabindex in the tree, just is the node focusable or not). So apparently some group (ARIA?) specified the behavior so that AT works well. However, I don't see it in [API Mappings 1.1](https://www.w3.org/TR/core-aam-1.1/#keyboard-focus_tabindex). In fact, that spec says
>Where tabindex="0", set the focusable state and include it in the sequential tab order.

Not sure what is going on since that is not what Firefox and Chrome do. Seems like the implementations do the right thing (and hence inheriting their behavior is good) and that the mapping spec needs a bug fix...

**Comment by faceless2** [comment]
I was about to post a new issue on this, but actually it ties in here.

Re. Brians post above above links having "all kinds of specialness" - a special aspect they need but that no-one has listed yet is alternative text to describe the _link purpose_ in order to meet accessibility criteria ([ref](https://www.w3.org/WAI/WCAG21/quickref/#link-purpose-in-context)).

The "alt" attribute in HTML covers this, but there isn't - yet - an equivalent for general MathML elements, just the "alttext" on the root `<math>` element.

I'm not a WCAG expert so while it seems clear to me that some sort of "alt" attribute is required, I'm prepared to be corrected on this. But we're outputting MathML to PDF and I'm much more familiar with the equivalent specifications there (chiefly ISO14289). This requires alternative text on all links, not just those where the "_purpose of each link can be determined from the link text alone_" (as described in WCAG). So to meet this criteria, some sort of "alt" attribute is essential.

**Comment by NSoiffer** [comment]
Your point about "alt" and accessibility is a good one. In general, it isn't needed because links on MathML elements would normally have an obvious purpose (a definition or clarification of the symbol/notation) and therefore would meet the accessibility criteria. However, it is possible that someone might use them for some other reason, so it makes sense to allow for "alt" to be used on the MathML elements that accept "alt".

I'm on the 14289 committee and will suggest amending the text of 14289-2 to be consistent with WCAG. Thanks for pointing that out! Unfortunately that doesn't help you now as 14289-2 is still a work in progress.
